File: tests/system/prom_helpers.py
"""
Helpers for parsing the Hydra.Core Prometheus metrics endpoint.

Extracted from `test_mix_precision_p_d_system.py` so the regex parsing
can be unit-tested without a live stack (the system tests need a running
Coordinator + llama-server, which CI doesn't have). See
`test_prom_helpers.py` for the unit tests.

Background: prometheus-net always renders labeled counters with their
label set, so `hydra_cross_model_kv_proceeded_total` shows up as

    hydra_cross_model_kv_proceeded_total{worker="rtx"} 1
    hydra_cross_model_kv_proceeded_total{worker="p100"} 0

A naive `^name\\s+...` regex (whitespace immediately after the name)
fails to match both — the gap before this fix was that the test asked
for the unlabeled total, got 0.0, and asserted a positive number, so
the test would have failed even when the guard was working correctly.

A correct reader must:
  1. Accept both labeled (`name{...}`) and unlabeled (`name `) forms.
  2. Sum across all label series when the caller asks for the total
     of a multi-label counter (or filter to a specific label set).
"""

from __future__ import annotations

import re
from typing import Mapping


# Anchor at the start of a line. The `(?:\s+|{)` allows either whitespace
# (unlabeled series) or `{` (labeled series) after the metric name. The
# value at the end is the standard Prometheus float grammar (with
# optional exponent) but we use a permissive `[0-9.eE+\-]+` so that
# NaN, Inf, and other special values parse as 0.0 in downstream code.
_LINE_PATTERN = re.compile(
    r"^(?P<name>[a-zA-Z_:][a-zA-Z0-9_:]*)"
    r"(?:\{(?P<labels>[^}]*)\})?"
    r"\s+(?P<value>[0-9.eE+\-]+)\s*$"
)


def parse_prom_lines(body: str) -> list[tuple[str, Mapping[str, str], float]]:
    """
    Parse a Prometheus text-exposition body into a list of
    (name, labels, value) tuples. Skip `# HELP` / `# TYPE` / blank
    lines silently.

    `labels` is an empty mapping for unlabeled series. The value is
    a float (NaN and Inf become 0.0 via `float()`'s parser — callers
    that need to distinguish can use `math.isnan` / `math.isinf` on
    the result).
    """
    out: list[tuple[str, Mapping[str, str], float]] = []
    for line in body.splitlines():
        if not line or line.startswith("#"):
            continue
        m = _LINE_PATTERN.match(line)
        if not m:
            continue
        name = m.group("name")
        labels_str = m.group("labels") or ""
        labels: dict[str, str] = {}
        if labels_str:
            for kv in labels_str.split(","):
                k, _, v = kv.partition("=")
                # Strip surrounding quotes from the value.
                v = v.strip()
                if v.startswith('"') and v.endswith('"'):
                    v = v[1:-1]
                labels[k.strip()] = v
        try:
            value = float(m.group("value"))
        except ValueError:
            value = 0.0
        out.append((name, labels, value))
    return out


def sum_counter(
    samples: list[tuple[str, Mapping[str, str], float]],
    name: str,
    labels: Mapping[str, str] | None = None,
) -> float:
    """
    Sum a counter across all series matching `name` (and optionally the
    label subset `labels`). Returns 0.0 when no series match.

    When `labels` is None, sums every series with the matching name
    regardless of labels — this is the "total across all workers"
    query that system tests use to assert a guard fired at least once.

    When `labels` is provided, filters to series whose label set is
    a superset of `labels` (Prometheus label matching semantics —
    `labels={worker="rtx"}` matches `{worker="rtx",extra="x"}` too,
    but for our counters the only label is `worker` so the
    distinction doesn't matter).
    """
    total = 0.0
    for sample_name, sample_labels, value in samples:
        if sample_name != name:
            continue
        if labels is not None and not all(
            sample_labels.get(k) == v for k, v in labels.items()
        ):
            continue
        total += value
    return total

File: src/core/Tests.Shared/PromHelpers.cs
using System.Globalization;
using System.Text.RegularExpressions;

namespace Tests.Shared;

/// <summary>
/// Helpers for parsing the Hydra.Core Prometheus text-exposition endpoint.
///
/// Ported from Python <c>tests/system/prom_helpers.py</c>. Extracted so the
/// regex parsing can be unit-tested without a live stack.
///
/// Background: prometheus-net always renders labeled counters with their
/// label set, so <c>hydra_cross_model_kv_proceeded_total</c> shows up as
/// <code>
///   hydra_cross_model_kv_proceeded_total{worker="rtx"} 1
///   hydra_cross_model_kv_proceeded_total{worker="p100"} 0
/// </code>
/// A naive <c>^name\\s+...</c> regex (whitespace immediately after the name)
/// fails to match both. A correct reader must:
///   1. Accept both labeled (<c>name{...}</c>) and unlabeled (<c>name </c>) forms.
///   2. Sum across all label series when the caller asks for the total.
/// </summary>
public static class PromHelpers
{
    /// <summary>
    /// A single parsed Prometheus sample.
    /// </summary>
    public sealed record Sample(
        string Name,
        IReadOnlyDictionary<string, string> Labels,
        double Value);

    // Anchored at start-of-line. The `(?:\{...\}|\\s+)` allows either
    // `{labels}` (labeled series) or whitespace (unlabeled series) after
    // the metric name. The value is the standard Prometheus float grammar
    // (with optional exponent) but we use a permissive `[0-9.eE+\-]+`
    // so that NaN, Inf, and other special values parse as 0.0.
    private static readonly Regex LinePattern = new(
        @"^(?<name>[a-zA-Z_:][a-zA-Z0-9_:]*)(?:\{(?<labels>[^}]*)\})?\s+(?<value>[0-9.eE+\-]+)\s*$",
        RegexOptions.Compiled);

    /// <summary>
    /// Parse a Prometheus text-exposition body into a list of
    /// <see cref="Sample"/> records. Skips blank lines and
    /// <c># HELP</c> / <c># TYPE</c> comment lines silently.
    ///
    /// Labels is an empty dictionary for unlabeled series. The value is
    /// a double (NaN and Inf become 0.0 via the permissive regex — callers
    /// that need to distinguish can check <see cref="double.IsNaN"/> /
    /// <see cref="double.IsInfinity"/> on the result).
    /// </summary>
    public static IReadOnlyList<Sample> ParsePromLines(string body)
    {
        var outList = new List<Sample>();

        foreach (var line in body.Split('\n'))
        {
            // Trim for robustness (Python splitlines also strips trailing \r).
            var trimmed = line.TrimEnd('\r');

            if (string.IsNullOrEmpty(trimmed) || trimmed.StartsWith('#'))
                continue;

            var m = LinePattern.Match(trimmed);
            if (!m.Success)
                continue;

            var name = m.Groups["name"].Value;
            var labelsStr = m.Groups["labels"].Value;
            var labels = new Dictionary<string, string>();

            if (!string.IsNullOrEmpty(labelsStr))
            {
                foreach (var kv in labelsStr.Split(','))
                {
                    var eqIdx = kv.IndexOf('=');
                    if (eqIdx < 0) continue;

                    var k = kv[..eqIdx].Trim();
                    var v = kv[(eqIdx + 1)..].Trim();

                    // Strip surrounding quotes from the value.
                    if (v.Length >= 2 && v[0] == '"' && v[^1] == '"')
                        v = v[1..^1];

                    labels[k] = v;
                }
            }

            double value = 0.0;
            if (double.TryParse(
                    m.Groups["value"].Value,
                    NumberStyles.Float,
                    CultureInfo.InvariantCulture,
                    out var parsed))
            {
                value = parsed;
            }

            outList.Add(new Sample(name, labels, value));
        }

        return outList;
    }

    /// <summary>
    /// Sum a counter across all series matching <paramref name="name"/>
    /// (and optionally the label subset <paramref name="labels"/>).
    /// Returns 0.0 when no series match.
    ///
    /// When <paramref name="labels"/> is <c>null</c>, sums every series
    /// with the matching name regardless of labels — this is the "total
    /// across all workers" query that system tests use to assert a guard
    /// fired at least once.
    ///
    /// When <paramref name="labels"/> is provided, filters to series whose
    /// label set is a superset of <paramref name="labels"/> (Prometheus
    /// label matching semantics).
    /// </summary>
    public static double SumCounter(
        IReadOnlyList<Sample> samples,
        string name,
        IReadOnlyDictionary<string, string>? labels = null)
    {
        var total = 0.0;

        foreach (var sample in samples)
        {
            if (sample.Name != name)
                continue;

            if (labels is not null)
            {
                var match = true;
                foreach (var kv in labels)
                {
                    if (!sample.Labels.TryGetValue(kv.Key, out var v) || v != kv.Value)
                    {
                        match = false;
                        break;
                    }
                }
                if (!match)
                    continue;
            }

            total += sample.Value;
        }

        return total;
    }
}
